### PR TITLE
chore(lib/common) implement byte pool to improve websocket subscription efficiency

### LINKED ISSUE #1477: Improve performance of websocket subscription using buffered pool
## Issue summary

- Currently, we are generating a random number for subscriptions ID to find an empty slot for a subscription. We might end up wasting too much time generating a random number because Intn doesn't generate a purely random number. 
- Instead, we can use a buffered pool to achieve the same.
- Buffered pool implemented using channels is also thread-safe. (Locks are not required)

## Other information and links

- We can modify this code to serve our use case https://github.com/oxtoacart/bpool/blob/master/bufferpool.go

### PR DESCRIPTION
**Pull request by edwardmack**
## Changes

- Issue #1477 outlines suggestions for changes to improve performance during websocket initialization.  This update introduces a byte pool that replaces the random number generator.
-
-

## Tests

```
go test ./lib/common/...
```

## Issues

- closes #1477

### CODE REVIEW

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]
I guess we should use `bs.importedBytepool.Put(byte)` in the `Unregister` function

File: dot/state/block.go [hunk 1]
```diff
@@ -46,10 +46,12 @@ type BlockState struct {
...
-	imported      map[byte]chan<- *types.Block
-	finalised     map[byte]chan<- *types.FinalisationInfo
-	importedLock  sync.RWMutex
-	finalisedLock sync.RWMutex
+	imported          map[byte]chan<- *types.Block
+	finalised         map[byte]chan<- *types.FinalisationInfo
+	importedLock      sync.RWMutex
+	finalisedLock     sync.RWMutex
+	importedBytepool  *common.BytePool
+	finalisedBytepool *common.BytePool
```
Comment (noot) [inline dot/state/block.go:54]:
```suggestion
	importedBytePool  *common.BytePool
	finalisedBytePool *common.BytePool
```

File: lib/common/bytepool.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	return
+}
+
+// Put returns the given Buffer to the BytePool.
+func (bp *BytePool) Put(b byte) error {
+	select {
+	case bp.c <- b:
+		return nil
+	default: // Discard the buffer if the pool is full.
```
Comment (noot) [inline lib/common/bytepool.go:58]: remove comment

File: dot/state/block.go (hunk 1 above)
Comment (edwardmack) [reply dot/state/block.go:54]: updated.

File: lib/common/bytepool.go (hunk 2 above)
Comment (edwardmack) [reply lib/common/bytepool.go:58]: comment removed.

**Comment by edwardmack** [comment]
> I guess we should use `bs.importedBytepool.Put(byte)` in the `Unregister` function

Good point, I've added this.

**Review by EclesioMeloJunior — APPROVED** [review]

File: lib/common/bytepool.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	select {
+	case bp.c <- b:
+		return nil
+	default:
+		return fmt.Errorf("pool is full")
+	}
+}
+
+// NumPooled returns the number of items currently pooled.
+func (bp *BytePool) NumPooled() int {
```
Comment (timwu20) [inline lib/common/bytepool.go:64]:
```suggestion
func (bp *BytePool) Len() int {
```

**Review by timwu20 — APPROVED** [review]

File: lib/common/bytepool.go (hunk 3 above)
Comment (edwardmack) [reply lib/common/bytepool.go]: updated.

**Pull request merged**